### PR TITLE
Make sure `kubectl cp` fails if source file doesn't exist in pod

### LINKED ISSUE #980: `kubectl cp` accidentally succeed if source file doesn't exist in pod
**What happened**:

`kubectl cp` accidentally succeed if source file doesn't exist in pod

**What you expected to happen**:

`kubectl cp` terminates with non-zero exit code and print error message.

**How to reproduce it (as minimally and precisely as possible)**:

1 `kubectl cp` now terminates with zero exit code even if source file doesn't exist in pod:
```sh
$ kubectl cp iperf-7d64f95bd5-dnmdv:/tmp/no-such-file /tmp/foo
tar: Removing leading `/' from member names
tar: /tmp/no-such-file: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
$ echo $?
0
```

2. `kubectl cp` sometimes even doesn't log any error:
```sh
$ cat reproduce.sh
#!/bin/bash

set -euo pipefail

for i in {1..10}; do
    out=$(kubectl cp iperf-7d64f95bd5-dnmdv:/tmp/no-such-file /tmp/foo)
    [[ $out =~ 'No such file or directory' ]] || break
    echo "$i"
    echo "$out"
done

$ ./reproduce.sh
1
tar: Removing leading `/' from member names
tar: /tmp/no-such-file: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
2
tar: Removing leading `/' from member names
tar: /tmp/no-such-file: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes client and server versions (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-12T01:09:16Z", GoVersion:"go1.15.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): macOS

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: knight42 <anonymousknight96@gmail.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

`kubectl cp` would spawn a separate goroutine to execute `tar` command in the container:
https://github.com/kubernetes/kubectl/blob/b155278f1f4a21a0be2d4f6f0037258dee4d1a22/pkg/cmd/cp/cp.go#L319-L323
If `tar` command fails, the main goroutine doesn't wait for that separate goroutine to finish and might exit with zero exit code.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/980

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubectl: make sure `kubectl cp` fails if source file doesn't exist in pod
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by soltysh — CHANGES_REQUESTED** [review]
Not quite sure what you're trying to achieve looking at the code, but based on the description I was expecting a code which will check the return code from the container and return error if such was reported.

**Comment by knight42** [comment]
> Not quite sure what you're trying to achieve

@soltysh I want to make sure `kubectl` exit with non-zero code and always print error if the source file doesn't exist in the pod 

> which will check the return code from the container and return error if such was reported.

It is unlikely to check the return code, because the `tar` command is executed in the container in the same way as we invoke `kubectl exec <pod name> -- tar ...` in `kubectl cp`:
https://github.com/kubernetes/kubernetes/blob/1df1f882c4421d313abb468fef5c15f89983471c/staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go#L516-L532

And if the `tar` command fails, `options.Run()` would return an error so we could simply propagate this error.

Now the problem is,  even if `tar` command fails, `o.untarAll` would still succeed and return no error, and there is chance that this line 
https://github.com/kubernetes/kubernetes/blob/1df1f882c4421d313abb468fef5c15f89983471c/staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go#L329
returns before the following line
https://github.com/kubernetes/kubernetes/blob/1df1f882c4421d313abb468fef5c15f89983471c/staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go#L322
could print an error, as they belongs to different goroutines.

**Comment by soltysh** [comment]
> could print an error, as they belongs to different goroutines.

we could try running exec and untar in separate goroutines and make the main thread wait for both, this way we'd be easily able to break each and get error.

**Comment by knight42** [comment]
> we could try running exec and untar in separate goroutines and make the main thread wait for both

@soltysh  IIUC this is similar to what I did in this PR. 

The only difference is I ran untar in the main goroutine and exec in a separate goroutine, and wait for the exec goroutine to exit before we return in the main goroutine:
https://github.com/kubernetes/kubernetes/pull/96566/files#diff-fc43d24429f5a0224892a2ac3f5b85d63b5b3808558b74b5bee083a08a22dd29R316-R335

**Comment by knight42** [comment]
@soltysh a friendly ping. I don't know if I have made myself clear, please let me know if there are issues I need to address.

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go [hunk 1]
```diff
@@ -293,40 +293,46 @@ func (o *CopyOptions) copyToPod(src, dest fileSpec, options *exec.ExecOptions) e
...
 	prefix = path.Clean(prefix)
 	// remove extraneous path shortcuts - these could occur if a path contained extra "../"
 	// and attempted to navigate beyond "/" in a remote filesystem
 	prefix = stripPathShortcuts(prefix)
-	return o.untarAll(src, reader, dest.File, prefix)
+	err := o.untarAll(src, reader, dest.File, prefix)
+	if err != nil {
+		return err
+	}
+	return <-errCh
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:335]: `errCh` is an unbuffered channel. Does this block if it never yields an error?

Comment (knight42) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:335]: Yes, and I think that is what we want. If source file doesn't exist in source Pod and we don't block here, the command might succeed accidentally.

Comment (eddiezane) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:335]: Ah I get it - you're relying on the defer at 318 to close that channel eventually.

Something might still be off though with the output.

```
 /tmp ~/Codez/kubernetes/kubernetes/_output/bin/kubectl cp nginx:/tmp/foo .
tar: Removing leading `/' from member names
tar: /tmp/foo: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
E0216 18:26:53.490999   80436 v2.go:147] io: read/write on closed pipe
command terminated with exit code 2
```

**Review by soltysh — CHANGES_REQUESTED** [review]
I looked at this again today, and I started wondering how hard it would be to try to verify if src exists in the tar during extraction? This would be the simplest and most straightforward way of fixing this issue, don't you think?

**Comment by gnarlex** [comment]
This is still an issue

/reopen /remove-lifecycle rotten

**Pull request closed without merging**